### DISCUSSION TITLE
Latest build (2022.314.0) runs noticeably slower.

### DISCUSSION THREAD

**Discussion by KoleckOLP** [ppy/osu#17291] [Q&A] [answered]
Yesterday I wanted to play more osu and I noticed my client got updated to 2022.314.0
When I go into a beatmap I noticed the game is running slower/laggier than before.

I usually check my draw response time when this happens. 
my draw response is usually 0.4ms - 0.5ms usually staying at ~0.47ms which feels great.
but in the latest build it's more like 0.5ms - 0.6ms at this point I usually check what is running in the background.

I turned off all the apps running in background including discord and my antivirus, I usually let those run when I play osu.
But it didn't help, and if the draw times seem not noticeable, they actually are I can see/feel that my cursor is laggier, approach circles are too.

I noticed the mouse cursor movement is slower even in the menu, so it's not only in gameplay.

I just can't play on the latest build.

Rebooting the PC and trying again didn't fix it still runs slower the previous build 2022.306.0 which still runs fine and smooth.
The issue with running an older build is that I get popups that "score couldn't be submitted"

I have also tried building Master from source and it's as slow as 2022.314.0

Anyone else noticed the latest build runs worse? Any fix to this?

**Comment by frenzibyte** [comment]
Since you already have the ability to build from `master`, try bisecting between `2022.314.0` and `2022.306.0` to figure out which commit specifically caused the slowness for you (see https://git-scm.com/docs/git-bisect).

**Reply by KoleckOLP** [reply]
Great idea thank you, I will do that as soon as I can!

**Reply by KoleckOLP** [reply]
Bisecting gave me these 3 commits:

8b3360d6d20d3d680b0fda0035ed5efab15a5834 (I skipped because it does not build)
ac55fea3c953a6c81326308200d4b905cd921c3b (I skipped because it does not build)
8fdf2b13ac2d83b206d88940e7582ae80c91a645 (this one has the slow/laggy behavior)

**Reply by peppy** [reply]
So if anything, it's related to borderless then. Did you try fullscreen yet..? Do you have two or more monitors? Potentially running at different refresh rates?

**Reply by KoleckOLP** [reply]
Yes I have two monitors both 1920x1080 one primary that I play osu on is 144Hz and the other is just 60Hz

Playing in exclusive fullscreen on the latest build feels as smooth as it did on the older build.

Here is a quick video comparing the two version but I can't tell the diference in the video, but I can tell it in the game. https://www.youtube.com/watch?v=Svg_wpPg0U8

**Reply by peppy** [reply]
Can you also take a screenshot of the layout of your displays (in the windows display settings)?

**Reply by KoleckOLP** [reply]
I will do that when I get home from work today.

**Comment by peppy** [comment]
A change of anything near or below 1ms is not visually noticeable, so let's try taking a step back. What are you seeing that feels "noticeably slower" to you? Are you able to record this using a camera or screen capture?

Reporting issues based on 0.1ms differences is nothing more than consuming developer time which could be spent elsewhere.

**Reply by KoleckOLP** [reply]
I was afraid the issue will be dismissed; I don't know what is causing the game to feel less smooth. I will try looking into it more.

**Reply by peppy** [reply]
If it just feels "laggy" then there's a high chance you aren't running exclusive fullscreen. There was a recent change made which could be causing issues with this (although it was tested). Try alt-tabbing and confirm that there is a flash as your monitor/driver switches away from exclusive fullscreen (or attempt to open windows game bar with win-g and ensrue it doesn't open but only flashes the screen once).

**Reply by KoleckOLP** [reply]
I am running in borderless fullscreen mode; the slowness will go away if I switch to exclusive fullscreen?
I will test that when I get home.
The game use to run smooth even in borderless.

**Reply by KoleckOLP** [reply]
If you do need to see anything else, please do let me know.

Number 2 is my primary monitor (even though it's 2) it's the 144Hz one
Number 1 is my secondary monitor

I also apologize my Windows is in my native language Czech.

![ApplicationFrameHost_l1EXfVZy14](https://user-images.githubusercontent.com/22729896/159032647-577667ec-5d92-4524-abc6-e38c6b455fe2.png)

**Reply by Susko3** [reply]
If you rearrange the monitors so that 1 is left and 2 is right, what performance do you get?

**Reply by KoleckOLP** [reply]
Oh yeah, the performance when I swap the monitors around in windows settings is good.

But I do not want to have my monitors set like that.

Can it be fixed? Do you have any clue what is causing it?

**Reply by Susko3** [reply]
If you run the old version in this monitor configuration, it should show performance issues.

In https://github.com/ppy/osu-framework/pull/5068 the window was changed from extending 1px up and 1px left to just extending 1px to the right. Unfortunately this means that the window will now bleed 1px over to the 60 Hz monitor and your system can't really handle it (even though nothing is rendered there, it's just a black line).

For fixes:
1. **run osu! in fullscreen mode** (which you should be doing anyways for best performance)
   - do you have a specific reason for not running in fullscreen?
2. disable fullscreen optimizations: ([link](https://devblogs.microsoft.com/directx/demystifying-full-screen-optimizations/) § Troubleshooting) - probably won't work
3. move the window one pixel to the left (you'll have to use external tools to do it)

**Reply by bdach** [reply]
I do wonder whether the hack can somehow detect screen layout and do the 1px trick in a direction where there is no other display...?

**Reply by KoleckOLP** [reply]
I was running in borderless for two reasons, if I change my system volume while in osu the volume indicator does not draw over the game in Exclusive Fullscreen, and I do want to set it to an exact number.

And the other reason is to be able to switch to discord on the second monitor faster.

I also just tried putting the game to windowed mode and using [BorderlessGaming](https://github.com/Codeusa/Borderless-Gaming) to Fullscreen it, and performance seems good doing it this way.

**Reply by Susko3** [reply]
Could probably be done, but it will require moving the GL viewport if we don't want one pixel to be cut off. And if we move the viewport, then things that are related to window coordinates (such as absolute mouse input) will be off-by-one. Accounting for things that will be off-by-one seems like unnecessary work...

Borderless isn't really meant for performance.

**Reply by Susko3** [reply]
> I also just tried putting the game to windowed mode and using [BorderlessGaming](https://github.com/Codeusa/Borderless-Gaming) to Fullscreen it, and performance seems good doing it this way.

If you take a screenshot (F12), is it missing one pixel? (eg. 1919x1080 instead of 1920x1080)

**Reply by KoleckOLP** [reply]
Nope the game is 1920x1080 and it even says so when BG activates.

![osu_2022-03-18_20-08-41](https://user-images.githubusercontent.com/22729896/159068741-5e627437-abc0-4c88-9329-fd90fb60c4a5.png)

**Reply by KoleckOLP** [reply]
It in game says 1920x1080

![osu_2022-03-18_20-10-16](https://user-images.githubusercontent.com/22729896/159069028-2202209c-c279-4db8-b6b6-802d18ad7cff.png)

**Reply by peppy** [reply]
> I was running in borderless for two reasons, if I change my system volume while in osu the volume indicator does not draw over the game in Exclusive Fullscreen, and I do want to set it to an exact number.
> 
> And the other reason is to be able to switch to discord on the second monitor faster.
> 
> I also just tried putting the game to windowed mode and using [BorderlessGaming](https://github.com/Codeusa/Borderless-Gaming) to Fullscreen it, and performance seems good doing it this way.

We should definitely look into how this is being done.

**Reply by frenzibyte** [reply]
Looks to boil down all the way to [`Manipulation.MakeWindowBorderless`](https://github.com/Codeusa/Borderless-Gaming/blob/74b19ecebc4bae4df1fbb1776ec7c5d69d4e0d0c/BorderlessGaming.Logic/Windows/Manipulation.cs#L31-L220).

**Reply by KoleckOLP** [reply]
It works very well on most games I tried it with. Would recommend.

**Reply by Susko3** [reply]
Tested BorderlessGaming; on my AMD R7 370 and Windows 10 21H1 it behaves the same as exclusive fullscreen, except that osu! doesn't minimize when not in focus and mouse confine mode not being forced on. With the usual problems such as the displays showing black for one frame when alt-tabbing. Same thing with fullscreen optimizations disabled.

@KoleckOLP how does this compare to you? Does your screen briefly show black when alt-tabbing? What are the differences between (exclusive) fullscreen mode in osu! and borderless fullscreen in BorderlessGaming?

**Reply by KoleckOLP** [reply]
I am on Windows 10 21H2, and my GPU is the Nvidia GeForce RTX 3070.

The game does not flicker at all when in windowed mode and fullscreened by BorderlessGaming.

In Exclusive Fullscreen it takes a while to switch, and I can't just move mouse of the window.

I play on a graphic tablet and it's set to only the primary screen, and when I want to go to the secondary Monitor, I use the mouse, no alt-tabbing.

I don't understand the question I already explained the issue, the issue was with performerce because of a pixel overlapping to 60Hz monitor,
And I want to play on borderless because of Windows Volume dialog not drawing over Exclusive Fullscreen and being able to mouse out of the game and take screenshots using ShareX.

**Reply by peppy** [reply]
I'm not sure if I already asked this, but can you confirm that disabling your second display also fixes the issue?

**Reply by KoleckOLP** [reply]
I will try it when I get back home and get back to you.

Other than that, I've been playing using Borderless Gaming and it works super well.

**Reply by KoleckOLP** [reply]
Ehh I can't tell if it helped or not but switching the display positions did.

**Comment by smoogipoo** [comment]
Draw time is irrelevant. You should open the frame time graphs by pressing Ctrl+F11 twice and seeing if there are any spikes, and taking a video of such spikes.

It could be intermittent stutters due to increased GC load in the last while that was addressed via https://github.com/ppy/osu/pull/17243, but it's hard to say without the aforementioned video.

**Reply by KoleckOLP** [reply]
I do not think it's spikes, I can see that the game is less smooth and it throws my aim off.
I will try to record it and checkout the frame time graphs.

**Comment by Susko3** [comment, marked as answer]
Same underlying cause as https://github.com/ppy/osu/issues/12535.
